target/linux/ramips/image/mt7621.mk

```diff
@@ -815,6 +815,19 @@ define Device/elecom_wrc-2533gst2
 endef
 TARGET_DEVICES += elecom_wrc-2533gst2
 
+define Device/etisalat_s3
+  $(Device/sercomm_dxx)
+  IMAGE_SIZE := 32768k
```

Comment:
It is recommended to add ` | check-size` here to limit image size. And the `IMAGE_SIZE` can be 0x60000(kernel) + 0x2000000(file system)? I'm not sure.
https://github.com/openwrt/openwrt/blob/aa12a0fdd1c5a004281633c5b0758da1781bb41c/target/linux/ramips/image/common-sercomm.mk#L82-L84

Reply:
> It is recommended to add ` | check-size` here to limit image size. And the `IMAGE_SIZE` can be 0x60000(kernel) + 0x2000000(file system)? I'm not sure.

Yes, it would be nice to do such a check. Let's think together about the limit:
1. 0x600000 is the max possible kernel size. Real kernel size can be much less. Current size - 2771591 bytes.
2. kernel(2771591) + rootfs(5505024) > factory(7123904) because we have gzip in the factory recipe.

So, I think it would be safer to keep IMAGE_SIZE == 0x2000000(file system) if the strict calculation couldn't be done.

Reply:
@DragonBluep `| check-size` has been added

Reply:
Yep, maybe 0x2000000 is a good value. I believe 32MiB is big enough for OpenWrt official build.

Reply:
@DragonBluep It seems I found a clean solution. lol. Just checking a pure ubi size.
```
  IMAGE/factory.img := append-ubi | check-size | \
	sercomm-part-tag rootfs | sercomm-prepend-tagged-kernel kernel | \
	gzip | sercomm-payload | sercomm-crypto | check-size
```
I'll push the commit and close this conversation.

Reply:
Yes, it should be a good solution for this device.

By the way. There are 20 blocks (2560 KiB) in the ubi that reserved for bad block management. Since the ubi partition of this device consists of several partitions, it just avoids the problem of computing the real available space. so `IMAGE_SIZE := 32768k` can be used to limit the max ubi size in factory image.
https://github.com/openwrt/openwrt/blob/aa12a0fdd1c5a004281633c5b0758da1781bb41c/target/linux/ramips/mt7621/config-5.15#L178

Reply:
> By the way. There are 20 blocks (2560 KiB) in the ubi that reserved for bad block management. Since the ubi partition of this device consists of several partitions, it just avoids the problem of computing the real available space. so `IMAGE_SIZE := 32768k` can be used to limit the max ubi size in factory image.

Thanks for the info. We should keep this in mind for devices with one ubi partition. Do I understand correctly that we should make a review of all devices and, if necessary, adjust IMAGE_SIZE?

P.S. `check-size` position fixed in https://github.com/openwrt/openwrt/commit/8d92a5482e20d2b89caa9b68981f498618b1ab96

Reply:
> Do I understand correctly that we should make a review of all devices and, if necessary, adjust IMAGE_SIZE?

Yes, but it's not very urgent now. NAND devices usually have enough space.

In addition, it seems that OpenWrt needs to reserve 2~3 blocks to save the configuration. This is a complex problem, and even NOR Flash devices are affected.